nntrainer/graph/network_graph.cpp

```diff
@@ -402,47 +449,61 @@ int NetworkGraph::realizeMultiOutputType(Layer &current) {
 }
 
 int NetworkGraph::isCompilable() {
-  if (layers.empty()) {
+  if (adj.empty()) {
     ml_loge("Layer is empty");
     return ML_ERROR_INVALID_PARAMETER;
   }
 
-  Layer &l = *layers[0];
-
-  /** Dimension of first layer must be known */
-  // TODO: move this to the input layer and not first layer
-  if (l.getInputDimension().size() == 0) {
-    ml_loge("InputDimension of first layer is not set");
-    return ML_ERROR_INVALID_PARAMETER;
-  }
+  return ML_ERROR_NONE;
+}
 
+int NetworkGraph::checkCompiledGraph() {
+  auto &l = Sorted[0].layer;
   /** First layer cannot be activation, batch normalization or loss */
-  const std::string &type = l.getType();
+  const std::string &type = l->getType();
   if (istrequal(type, ActivationLayer::type) ||
       istrequal(type, BatchNormalizationLayer::type) ||
       istrequal(type, LossLayer::type)) {
-    ml_loge("%s cannot be the first layer, type: %s", l.getName().c_str(),
+    ml_loge("%s cannot be the first layer, type: %s", l->getName().c_str(),
             type.c_str());
     return ML_ERROR_INVALID_PARAMETER;
   }
 
+  /** Dimension of input layers must be known */
+  for (auto const &lnode : Sorted) {
+    if (lnode.layer->getType() == InputLayer::type) {
+      if (lnode.layer->getInputDimension().size() == 0) {
+        ml_loge("InputDimension of first layer is not set");
+        return ML_ERROR_INVALID_PARAMETER;
+      }
+    }
+  }
+
   return ML_ERROR_NONE;
 }
 
-int NetworkGraph::setGraphNode(const LossType loss_type) {
+int NetworkGraph::setGraphNode() {
```

Comment:
is it ok if we remove the consideration of the loss layer?

Reply:
Loss layer can't be added till we know where to add it. Earlier code that the order of layers in the ini is sensible but that can be wrong. So, first, the graph is created without loss and sorted. After sorting, the graph knows the last layer, and the loss layer is inserted at the end.